### PR TITLE
CustomDataFrameAction Expectations

### PR DESCRIPTION
**Pull request by zzeekk**
see #882

### CODE REVIEW

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/CustomDataFrameActionTest.scala [hunk 1]
```diff
@@ -399,6 +401,95 @@ class CustomDataFrameActionTest extends FunSuite with BeforeAndAfter {
...
+
+    // prepare & start load
+    val customTransformerConfig1 = SQLDfsTransformer(code = Map(tgtDO1.id.id -> "select * from src1"))
+    val customTransformerConfig2 = SQLDfsTransformer(code = Map(tgtDO2.id.id -> "select * from src2"))
+    val action1 = CustomDataFrameAction("ca", List(srcDO1.id, srcDO2.id), List(tgtDO1.id, tgtDO2.id), mainInputId = Some(srcDO1.id), mainOutputId = Some(tgtDO1.id),
+      transformers = Seq(customTransformerConfig1, customTransformerConfig2),
+      expectations = Seq(TransferRateExpectation(), CompletnessExpectation(expectation = None))
+    )
+    instanceRegistry.register(action1)
+    val l = Seq(("jonson","rob",5),("doe","bob",3)).toDF("lastname", "firstname", "rating")
```
Comment (Geheiner) [inline sdl-core/src/test/scala/io/smartdatalake/workflow/action/CustomDataFrameActionTest.scala]: can you rename this? Maybe inputData?

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/CustomDataFrameActionTest.scala [hunk 2]
```diff
@@ -399,6 +401,95 @@ class CustomDataFrameActionTest extends FunSuite with BeforeAndAfter {
...
```
Comment (Geheiner) [inline sdl-core/src/test/scala/io/smartdatalake/workflow/action/CustomDataFrameActionTest.scala]: same here, rename to inputData

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/ExpectationValidation.scala [hunk 3]
```diff
@@ -178,7 +180,7 @@ private[smartdatalake] trait ExpectationValidation { this: DataObject with Smart
       })
     // throw exception on error, but log metrics before
     val errors = validationResults.filterKeys(_.failedSeverity == ExpectationSeverity.Error)
-    if (errors.nonEmpty) logger.error(s"($id) Expectation validation failed with metrics "+updatedMetrics.map{case(k,v) => s"$k=$v"}.mkString(" "))
+    if (errors.nonEmpty) logger.error(s"($id) Expectation validation failed with metrics "+orderMetrics(updatedMetrics).map{case(k,v) => s"$k=$v"}.mkString(" "))
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/ExpectationValidation.scala]: formatting

**Review by Geheiner — APPROVED** [review]

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/CustomDataFrameActionTest.scala (hunk 1 above)
Comment (zzeekk) [reply sdl-core/src/test/scala/io/smartdatalake/workflow/action/CustomDataFrameActionTest.scala]: Ok, done

File: sdl-core/src/test/scala/io/smartdatalake/workflow/action/CustomDataFrameActionTest.scala (hunk 2 above)
Comment (zzeekk) [reply sdl-core/src/test/scala/io/smartdatalake/workflow/action/CustomDataFrameActionTest.scala]: Ok, done

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/ExpectationValidation.scala (hunk 3 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/ExpectationValidation.scala]: Seems IntelliJ formatting on commit is only concerned about the changed lines.
I reformatted both files now...

**Pull request merged**